failure in udrcx/test_ucp_peer_failure_zcopy_multi.status_after_error/2

http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5435/label=hpc-test-node2,worker=2/console

```
19:19:09 [ RUN      ] udrcx/test_ucp_peer_failure_zcopy_multi.status_after_error/2
19:19:10 [hpc-test-node2:50186:0]   wireup_ep.c:331  Assertion `ucs_queue_is_empty(&self->pending_q)' failed
19:19:11 
19:19:11 /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/ucp/wireup/wireup_ep.c: [ ucp_wireup_ep_t_cleanup() ]
19:19:11       ...
19:19:11       328     ucp_worker_h worker = ucp_ep->worker;
19:19:11       329 
19:19:11       330     ucs_assert(ucs_queue_is_empty(&self->pending_q));
19:19:11 ==>   331     ucs_assert(self->pending_count == 0);
19:19:11       332 
19:19:11       333     ucs_debug("ep %p: destroy wireup ep %p", ucp_ep, self);
19:19:11       334 
19:19:11 
19:19:11 ==== backtrace ====
19:19:11  0 0x00000000000562be ucp_wireup_ep_t_cleanup()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/ucp/wireup/wireup_ep.c:331
19:19:11  1 0x0000000000058856 ucs_class_call_cleanup_chain()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/ucs/type/class.c:50
19:19:11  2 0x0000000000056030 ucp_wireup_ep_t_delete()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/ucp/wireup/wireup_ep.c:22
19:19:11  3 0x0000000000020a6b ucp_worker_iface_error_handler()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/ucp/core/ucp_worker.c:356
19:19:11  4 0x000000000006edf4 uct_ud_iface_dispatch_async_comps_do()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/uct/ib/ud/base/ud_iface.c:763
19:19:11  5 0x000000000007ad58 uct_ud_iface_dispatch_zcopy_comps()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/uct/ib/ud/base/ud_iface.h:440
19:19:11  6 0x000000000002210a ucs_callbackq_dispatch()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../src/ucs/datastruct/callbackq.h:168
19:19:11  7 0x000000000063dadf ucp_test_base::entity::progress()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/ucp/ucp_test.cc:441
19:19:11  8 0x000000000056bf92 test_ucp_peer_failure::wait_err()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/ucp/test_ucp_peer_failure.cc:109
19:19:11  9 0x000000000046fe76 ucs::test_base::run()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/test.cc:249
19:19:11 10 0x0000000000467493 testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/gtest-all.cc:3562
19:19:11 11 0x000000000045b90d testing::Test::Run()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/gtest-all.cc:3635
19:19:11 12 0x000000000045b9dc testing::TestInfo::Run()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/gtest-all.cc:3812
19:19:11 13 0x000000000045bb3f testing::TestCase::Run()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/gtest-all.cc:3930
19:19:11 14 0x00000000004604d7 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/gtest-all.cc:5802
19:19:11 15 0x00000000004607db testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/gtest-all.cc:5719
19:19:11 16 0x000000000040f1e3 main()  /scrap/jenkins/workspace/hpc-ucx-pr-2/label/hpc-test-node2/worker/1/contrib/../test/gtest/common/gtest.h:20059
19:19:11 17 0x0000000000021c05 __libc_start_main()  ???:0
19:19:11 18 0x0000000000446098 _start()  ???:0
19:19:11 ===================
19:19:11 Sending notification to yosefe@mellanox.com
19:19:14 [hpc-test-node2:50186:0] Process frozen...
21:58:45 
make: *** [test] Terminated
```

--- a/src/ucp/core/ucp_worker.c
+++ b/src/ucp/core/ucp_worker.c
@@ -318,18 +318,16 @@ ucp_worker_iface_error_handler(void *arg, uct_ep_h uct_ep, ucs_status_t status)
     uct_ep_h                aux_ep           = NULL;
     uct_tl_resource_desc_t* tl_rsc;
     uint64_t                dest_uuid UCS_V_UNUSED;
-    ucp_ep_h                ucp_ep_iter;
     khiter_t                ucp_ep_errh_iter;
     ucp_err_handler_cb_t    err_cb;
-    ucp_lane_index_t        lane, n_lanes, failed_lane;
+    ucp_lane_index_t        lane, failed_lane;
     ucp_rsc_index_t         rsc_index;
 
     /* TODO: need to optimize uct_ep -> ucp_ep lookup */
-    kh_foreach(&worker->ep_hash, dest_uuid, ucp_ep_iter, {
-        for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep_iter); ++lane) {
-            if ((uct_ep == ucp_ep_iter->uct_eps[lane]) ||
-                ucp_wireup_ep_test_aux(ucp_ep_iter->uct_eps[lane], uct_ep)) {
-                ucp_ep = ucp_ep_iter;
+    kh_foreach(&worker->ep_hash, dest_uuid, ucp_ep, {
+        for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+            if ((uct_ep == ucp_ep->uct_eps[lane]) ||
+                ucp_wireup_ep_test_aux(ucp_ep->uct_eps[lane], uct_ep)) {
                 failed_lane = lane;
                 goto found_ucp_ep;
             }
@@ -345,20 +343,27 @@ found_ucp_ep:
     rsc_index   = ucp_ep_get_rsc_index(ucp_ep, lane);
     tl_rsc      = &worker->context->tl_rscs[rsc_index].tl_rsc;
 
-    /* Purge outstanding */
-    uct_ep_pending_purge(ucp_ep_iter->uct_eps[lane], ucp_ep_err_pending_purge,
-                         UCS_STATUS_PTR(status));
+    /* Destroy all lanes except failed one since ucp_ep becomes unusable as well */
+    for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
+        if (ucp_ep->uct_eps[lane] == NULL) {
+            continue;
+        }
 
-    /* Destroy all lanes excepting failed one since ucp_ep becomes unusable as well */
-    for (lane = 0, n_lanes = ucp_ep_num_lanes(ucp_ep); lane < n_lanes; ++lane) {
-        if ((lane != failed_lane) && ucp_ep->uct_eps[lane]) {
-            /* TODO flush and purge outstanding operations */
+        /* Purge pending queue */
+        ucs_trace("ep %p: purge pending on uct_ep[%d]=%p", ucp_ep, lane,
+                  ucp_ep->uct_eps[lane]);
+        uct_ep_pending_purge(ucp_ep->uct_eps[lane], ucp_ep_err_pending_purge,
+                             UCS_STATUS_PTR(status));
+
+        if (lane != failed_lane) {
+            ucs_trace("ep %p: destroy uct_ep[%d]=%p", ucp_ep, lane,
+                      ucp_ep->uct_eps[lane]);
             uct_ep_destroy(ucp_ep->uct_eps[lane]);
             ucp_ep->uct_eps[lane] = NULL;
         }
     }
 
-    /* Move failed lane at 0 index */
+    /* Move failed lane to index 0 */
     if (failed_lane != 0) {
         ucp_ep->uct_eps[0] = ucp_ep->uct_eps[failed_lane];
         ucp_ep->uct_eps[failed_lane] = NULL;
@@ -367,10 +372,11 @@ found_ucp_ep:
     /* NOTE: if failed ep is wireup auxiliary then we need to replace the lane
      *       with failed ep and destroy wireup ep
      */
-    if (ucp_wireup_ep_test_aux(ucp_ep_iter->uct_eps[0], uct_ep)) {
-        aux_ep = ucp_wireup_ep_extract_aux(ucp_ep_iter->uct_eps[0]);
-        uct_ep_destroy(ucp_ep_iter->uct_eps[0]);
-        ucp_ep_iter->uct_eps[0] = aux_ep;
+    if (ucp_wireup_ep_test_aux(ucp_ep->uct_eps[0], uct_ep)) {
+        aux_ep = ucp_wireup_ep_extract_aux(ucp_ep->uct_eps[0]);
+        ucs_trace("ep %p: destroy failed uct_ep=%p", ucp_ep, ucp_ep->uct_eps[0]);
+        uct_ep_destroy(ucp_ep->uct_eps[0]);
+        ucp_ep->uct_eps[0] = aux_ep;
     }
 
     /* Redirect all lanes to failed one */

--- a/src/ucp/wireup/wireup_ep.c
+++ b/src/ucp/wireup/wireup_ep.c
@@ -192,12 +192,12 @@ ucp_wireup_ep_pending_purge(uct_ep_h uct_ep, uct_pending_purge_callback_t cb,
     uct_pending_req_t *req;
     ucp_request_t *ucp_req;
 
-    if (wireup_ep->aux_ep != NULL) {
-        ucs_queue_for_each_extract(req, &wireup_ep->pending_q, priv, 1) {
-            ucp_req = ucs_container_of(req, ucp_request_t, send.uct);
-            cb(&ucp_req->send.uct, arg);
-        }
+    ucs_queue_for_each_extract(req, &wireup_ep->pending_q, priv, 1) {
+        ucp_req = ucs_container_of(req, ucp_request_t, send.uct);
+        cb(&ucp_req->send.uct, arg);
+    }
 
+    if (wireup_ep->aux_ep != NULL) {
         uct_ep_pending_purge(wireup_ep->aux_ep, ucp_wireup_ep_pending_req_release,
                              arg);
     }

